### ISSUE TITLE
Codegen for AutogradNestedTensor

### ISSUE THREAD

**Issue by YifanShenSZ** [pytorch/pytorch#83041]
We are considering letting dispatch key `AutogradNestedTensor` to be nested-tensor-specific `CompositeImplicitAutograd`, in order to implement composite nested-tensor ops that cannot reuse the composite regular-tensor code.

To do that, we have to modify codegen.

**Comment by YifanShenSZ** [comment]
First error I ran in to is when I try to add nested-tensor `reshape_as`:
```
- func: reshape_as(Tensor(a) self, Tensor other) -> Tensor(a)
  variants: method
  device_check: NoCheck
  device_guard: False
  dispatch:
    CompositeImplicitAutograd: reshape_as
    AutogradNestedTensor: reshape_as_nested
```

Then I got:
```
Traceback (most recent call last):
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/setup_helpers/generate_code.py", line 232, in <module>
    main()
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/setup_helpers/generate_code.py", line 183, in main
    generate_code(
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/setup_helpers/generate_code.py", line 58, in generate_code
    gen_autograd(
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_autograd.py", line 84, in gen_autograd
    gen_inplace_or_view_type(
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 579, in gen_inplace_or_view_type
    fm.write_sharded(
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/torchgen/utils.py", line 265, in write_sharded
    env = env_callable(item)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 549, in gen_inplace_or_view_type_env
    definition = inplace_or_view_method_definition(fn)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/context.py", line 16, in wrapper
    return func(f)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 520, in inplace_or_view_method_definition
    type_definition_body=emit_inplace_or_view_body(fn),
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/context.py", line 16, in wrapper
    return func(f)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 477, in emit_inplace_or_view_body
    call, rhs_value = emit_view_body(fn, TMP_VAR)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 424, in emit_view_body
    call += emit_view_lambda(f, unpacked_bindings)
  File "/Users/yifanshensz/Software/Mine/pytorch-dev1/tools/autograd/gen_inplace_or_view_type.py", line 335, in emit_view_lambda
    raise TypeError(
TypeError: You are adding an ConstRefCType(elem=BaseCType(type=BaseCppType(ns='at', name='Tensor'))) other_ argument to op reshape_as in addition to known types: BaseCType(type=BaseCppType(ns='', name='int64_t')), OptionalCType(elem=BaseCType(type=BaseCppType(ns='', name='int64_t'))), BaseCType(type=BaseCppType(ns='', name='bool')), BaseCType(type=BaseCppType(ns='at', name='IntArrayRef')), BaseCType(type=BaseCppType(ns='c10', name='SymIntArrayRef')). Please update the list or materialize it so that it can be closed over by value, also add a test in pytorch/xla/test/test_operations.py where this code is exercised.
  in native_functions.yaml line aten/src/ATen/native/native_functions.yaml:4086:
    reshape_as(Tensor(a) self, Tensor other) -> Tensor(a)
  in native_functions.yaml line aten/src/ATen/native/native_functions.yaml:4086:
    reshape_as(Tensor(a) self, Tensor other) -> Tensor(a)
```

If we look at [gen_inplace_or_view_type.py](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_inplace_or_view_type.py#L321-L327), then we find that the regular-tensor `reshape_as` should also raise this error. The only reason why it wouldn't must be that `CompositeImplicitAutograd` never calls into this `emit_view_lamda` function. We should let `AutogradNestedTensor` to do the same.

**Comment by YifanShenSZ** [comment]
closed with [#82754](https://github.com/pytorch/pytorch/pull/82754) and [#84154](https://github.com/pytorch/pytorch/pull/84154)
